### DISCUSSION TITLE
Master Server Changelog Size

### DISCUSSION THREAD

**Discussion by anon314159** [moosefs/moosefs#612] [General]
Hello, 

Is there any documentation that describes what the change log files are located under /var/lib/mfs and what would be considered normal sizes relative to the number of metadata objects?  The original documentation said something like 300 bytes of overhead per object, but I don't know if this is actually correct.  Whereas, I have a 4 node replica 2 setup that consists of 28 million chunks, but there's a single change log file on the master server that's well over 300GB, which seems a bit ludacris relative to the number of chunks. As a total space for all chunks combined for this small is cluster is 160TB.

**Comment by chogata** [comment]
The changelog files' sizes are not dependent on the number of objects, but on the number of operations performed in your system. Basically one operation is one line in a changelog file. The more objects in a filesystem, the more (usually) operations, but this is not always true, there can be large, but quite static instances (for example for backup purposes), which generate small changelog files and not so large, but very active instances (databases, e-mails etc.), that generate large changelog files.

300 bytes per object is a rough estimate for RAM usage per object in master server - it has nothing to do with changelogs. More with the metadata file size.

In general: metadata file size (metadata.mfs, metadata.mfs.back etc.) depends on the RAM usage of your master. Changelog (changelog.*.mfs) file size depends on number of operations per hour in your system (both I/O and internal replication operations).

**Reply by anon314159** [reply]
Ah okay, thank you for the response.  Correct me if I am wrong but based on your description, changelogs behave more like a journal?  Are they circular or do they just continue growing indefinitely based on the number of operations over a given period of time?  Whats the impact of recycling the MooseFS master service on the changelog files? I noticed that the metaloggers also periodically sync these files from the master and depending on there size and available network bandwidth can take a considerable amount of time.

**Reply by chogata** [reply]
Yes, they are kind of a journal. They are used if your master process for some reason does not exit cleanly and does not write the metadata file upon exit or the latest metadata file gets corrupted for some reason. Then, the previous metadata file is read and then the changelogs, to update the previous metadata file to current state.

The changelogs are recycled. 

Check out your mfsmaster.cfg file for these variables: BACK_LOGS, BACK_META_KEEP_PREVIOUS and METADATA_SAVE_FREQ.

BACK_LOGS - tells you how many 1-hour changelogs are kept; default is 50, so you will have 50 files, each containing 1 hour of operations; once your master works for longer than 50 hours, the oldest file will be deleted to make space for the newest one.
BACK_META_KEEP_PREVIOUS - tells you how many old metadata file saves are kept; default is 1
METADATA_SAVE_FREQ - tells you how often your master saves the metadata file to disk; default is once a day, so every 24 hours

We always recommend to have enough changelogs to last beyond the oldest metadata file and with the defaults you have exactly that: you will always have 2 metadata files (the last one and 1 previous), so just a minute before the scheduled time of next metadata save your previous (not latest) metadata file will be almost 48 hours old, so if you are forced to use it, you need 48 hours of changelogs, which, with the default of 50, you will have :)
